### PR TITLE
Add device support for TS110E Model QS-Zigbee-D02-TRIAC-LN

### LINKED ISSUE #1415: [Device Support Request] Device support for Tuya TS110E
I am not sure if this is a device request, or if the device itself is just not working. Maybe someone can halp?

I ordered two of these in-wall dimmers:

* [blakadder.com entry](https://zigbee.blakadder.com/Lonsonho_QS-Zigbee-D02-TRIAC-LN.html)
* [Device on aliexpress.com](https://www.aliexpress.com/item/4001279149071.html?spm=a2g0o.9042311.0.0.de834c4d9jmLG4)

If successfully installed it and paired it with home assistant (I'm using zha) and was able to use the on-off switch right away, but the dimmer slider wouldn't do anything. So naturally I thought I'd need a quirk. However, I wasn't able to get the dimmer slider working.

I pasted the device signature below and you can see that the `in_clusters` already have `"0x0006"` and `"0x0008"` which is [`OnOff`](https://github.com/zigpy/zigpy/blob/fc8e923e2f2482810b9643c0b2bde149a56a1930/zigpy/zcl/clusters/general.py#L385) and [`LevelControl`](https://github.com/zigpy/zigpy/blob/fc8e923e2f2482810b9643c0b2bde149a56a1930/zigpy/zcl/clusters/general.py#L426). A lot of the other tuya quirks for the dimmers didn't have those, so I understand that a quirk was necessary. In this case however I was surprised that it wasn't working right out of the box, because the [blakadder.com entry](https://zigbee.blakadder.com/Lonsonho_QS-Zigbee-D02-TRIAC-LN.html) confirmed it working with ZHA. However, the blakadder.com entry lists it as `Zigbee ID: "TS110F"; "_TYZB01_qezuin6k"`, while the one I actually got it `Zigbee ID: "TS110E"; "_TZ3210_ngqk6jia"`. So a small difference, even though the device itself looks exactly as on the pictures on blakadder.com and aliexpress.com.

I came up with the following quirk that I stored in `tuya/ts110e.py`:
```python
class DimmerSwitchWithNeutral1Gang(TuyaDimmerSwitch):
    """Tuya Dimmer Switch Module With Neutral 1 Gang"""

    signature = {
        MODELS_INFO: [("_TZ3210_ngqk6jia", "TS110E")],
        ENDPOINTS: {
            # <SimpleDescriptor endpoint=1 profile=260 device_type=257
            # input_clusters=[0, 4, 5, 6, 8, 57345]
            # output_clusters=[10, 25]>
            1: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: zha.DeviceType.DIMMABLE_LIGHT,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    Groups.cluster_id,
                    Scenes.cluster_id,
                    OnOff.cluster_id,
                    LevelControl.cluster_id,
                    TuyaZBExternalSwitchTypeCluster.cluster_id,
                ],
                OUTPUT_CLUSTERS: [Time.cluster_id, Ota.cluster_id],
            },
            242: {
                # <SimpleDescriptor endpoint=242 profile=41440 device_type=97
                # input_clusters=[]
                # output_clusters=[33]
                PROFILE_ID: 41440,
                DEVICE_TYPE: 97,
                INPUT_CLUSTERS: [],
                OUTPUT_CLUSTERS: [GreenPowerProxy.cluster_id],
            },
        },
    }
    replacement = {
        ENDPOINTS: {
            1: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: zha.DeviceType.DIMMABLE_LIGHT,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    Groups.cluster_id,
                    Scenes.cluster_id,
                    OnOff.cluster_id,
                    LevelControl.cluster_id,
                    # TuyaManufacturerClusterOnOff,
                    # TuyaOnOff,
                    # TuyaManufacturerLevelControl,
                    # TuyaLevelControl,
                    TuyaZBExternalSwitchTypeCluster.cluster_id,
                ],
                OUTPUT_CLUSTERS: [Time.cluster_id, Ota.cluster_id],
            },
            242: {
                PROFILE_ID: 41440,
                DEVICE_TYPE: 97,
                INPUT_CLUSTERS: [],
                OUTPUT_CLUSTERS: [GreenPowerProxy.cluster_id],
            },
        },
    }
```

The quirk is seen by home assistant (I see the `Quirk: zhaquirks.tuya.ts110e.DimmerSwitchWithNeutral1Gang` on the device in the home assistant UI). As you can see I experimented with all combinations of the `OnOff` and `LevelControl` alternatives. So far only the original 
```
                    OnOff.cluster_id,
                    LevelControl.cluster_id,
```
works for at least the switch but not the dimmer.  
Other combinations i tried:
```
                    TuyaOnOff,
                    TuyaLevelControl,
```
that comes with error
```
2022-03-06 22:48:32 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [548252938000] 'DimmerSwitchWithNeutral1Gang' object has no attribute 'command_bus'
```
Which makes sense. So I tried
```
                    TuyaManufacturerClusterOnOff,
                    TuyaOnOff,
                    TuyaManufacturerLevelControl,
                    TuyaLevelControl,
```
Which didn't throw an error and allowed me to call the switch in the UI of home assistant, but it wouldn't turn on the light and the switch in the UI jumps back like 2 seconds later.

Does anyone have any other suggestions of what I could try?

# Device signature:
```json
{
  "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.Router: 1>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.AllocateAddress|RxOnWhenIdle|MainsPowered|FullFunctionDevice: 142>, manufacturer_code=4417, maximum_buffer_size=66, maximum_incoming_transfer_size=66, server_mask=10752, maximum_outgoing_transfer_size=66, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=False, *is_full_function_device=True, *is_mains_powered=True, *is_receiver_on_when_idle=True, *is_router=True, *is_security_capable=False)",
  "endpoints": {
    "1": {
      "profile_id": 260,
      "device_type": "0x0101",
      "in_clusters": [
        "0x0000",
        "0x0004",
        "0x0005",
        "0x0006",
        "0x0008",
        "0xe001"
      ],
      "out_clusters": [
        "0x000a",
        "0x0019"
      ]
    },
    "242": {
      "profile_id": 41440,
      "device_type": "0x0061",
      "in_clusters": [],
      "out_clusters": [
        "0x0021"
      ]
    }
  },
  "manufacturer": "_TZ3210_ngqk6jia",
  "model": "TS110E",
  "class": "zhaquirks.tuya.ts110e.DimmerSwitchWithNeutral1Gang"
}
```

**Additional context**

### PR DESCRIPTION
**Pull request by jonnylangefeld**
This device uses a non-default set level command.

This device looks exactly the same as TS110F but seems to have different zigbee command IDs. This is TS110F on blakadder: https://zigbee.blakadder.com/Lonsonho_QS-Zigbee-D02-TRIAC-LN.html

Tested in local docker environment.
Thanks to @javicalle this was possible.
Fix #1415

### CODE REVIEW

File: zhaquirks/tuya/ts110e.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            1: {
+                PROFILE_ID: zha.PROFILE_ID,
+                DEVICE_TYPE: zha.DeviceType.DIMMABLE_LIGHT,
+                INPUT_CLUSTERS: [
+                    Basic.cluster_id,
+                    Groups.cluster_id,
+                    Scenes.cluster_id,
+                    OnOff.cluster_id,
+                    F000LevelControlCluster,
+                    TuyaZBExternalSwitchTypeCluster.cluster_id,
```
Comment (javicalle) [inline zhaquirks/tuya/ts110e.py:131]: remove the '.cluster_id' part:
```python
                    TuyaZBExternalSwitchTypeCluster,
```

File: zhaquirks/tuya/ts110e.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+TUYA_LEVEL_ATTRIBUTE = 0xF000
+TUYA_CUSTOM_LEVEL_COMMAND = 0x00F0
+
+
+class TuyaLevelPayload(t.Struct):
+    """Tuya Level payload."""
+
+    level: t.uint16_t
+    transtime: t.uint16_t
```
Comment (javicalle) [inline zhaquirks/tuya/ts110e.py:36]: It should probably move to `tuya.__init__.py`, but let's keep it here for now.

**Review by javicalle — CHANGES_REQUESTED** [review]
A little fix for the checks

File: zhaquirks/tuya/ts110e.py (hunk 1 above)
Comment (MattWestb) [reply zhaquirks/tuya/ts110e.py:131]: I think first testing if the cluster / attribute is working on the device or not and is its not working then deleting it in the replacement (its always possible adding it later if its needed) but its very likely not working on this device.

The implementation is here:
https://github.com/zigpy/zha-device-handlers/blob/8ae597d335e2f19f73fdc82f78fb5ae5fb18a40c/zhaquirks/tuya/__init__.py#L887-L902

File: zhaquirks/tuya/ts110e.py (hunk 2 above)
Comment (MattWestb) [reply zhaquirks/tuya/ts110e.py:36]: Or first do the implementation in the quirk and getting it merged  and its working OK and then making one more PR for moving it to tuya INIT then tuya INIT is little tricky with many addings and can needing rebasing many times and its PITA (i have one PR waiting for being merged to it).

Comment (javicalle) [reply zhaquirks/tuya/ts110e.py:36]: Yes, that's the idea.
Keep here and merge. When required, move to Tuya INIT. 👍🏻

File: zhaquirks/tuya/ts110e.py (hunk 1 above)
Comment (jonnylangefeld) [reply zhaquirks/tuya/ts110e.py]: I actually noticed that I had to leave in the `.cluster_id` part, otherwise the quirk wouldn't show up on the device anymore.

File: zhaquirks/tuya/ts110e.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+class TuyaBulbType(t.enum8):
+    """Tuya bulb type."""
+
+    LED = 0x00
+    INCANDESCENT = 0x01
+    HALOGEN = 0x02
+
+
+class F000LevelControlCluster(LevelControl):
```
Comment (jonnylangefeld) [inline zhaquirks/tuya/ts110e.py:52]: This `LevelControl` part is actually making issues. Because it applies to all dimmers in my house now, even the non-Tuya ones. And those obviously don't need this quirk. I think I somehow need to incorporate `CustomCluster` somewhere. But just adding `class F000LevelControlCluster(CustomCluster, LevelControl):` doesn't work.
@javicalle do you know how I can change this quirk to only work with the Custom device `DimmerSwitchWithNeutral1Gang` below?

Comment (javicalle) [reply zhaquirks/tuya/ts110e.py:52]: Not sure what you mean. The `` cluster definition is only used in the `DimmerSwitchWithNeutral1Gang` quirk. An this quirk only applies to `_TZ3210_ngqk6jia` device.

What is exactly the problem? When they started? Have you updated HA?

If you believe that there's a problem with your quirk, try to remove it an check if the problem persist, but I don't know how that quirk can affect other devices than the `_TZ3210_ngqk6jia`.

Comment (jonnylangefeld) [reply zhaquirks/tuya/ts110e.py:52]: So the `zhaquirks.tuya.ts110e.DimmerSwitchWithNeutral1Gang` quirk didn't show up on the new dimmer (non tuya) I bought, but the `F000LevelControlCluster` showed up on the new dimmer I bought. I found that that is because we just used `class F000LevelControlCluster(LevelControl):`, without implementing a `CustomCluster`.
I have experimented some more and found that adding `CustomCluster` as mentioned in my last comment didn't work, because that sent the manufacturer ID to the device and the device didn't recognize that. I found out by continuously comparing the logs between the setting that worked and the one that didn't work.

Long story short, I had to add `NoManufacturerCluster` to make it work. That is only in the later versions of `zha-device-handlers`, so I had to update my entire home-assistance instance. I have done that now and updated my quirk to use `class F000LevelControlCluster(NoManufacturerCluster, LevelControl):`. And that ended up doing what I wanted which is that the `F000LevelControlCluster` only applies to the tuya dimmer and not to others.

I will update this PR accordingly as there are some other changes with the latest updates on the dev branch.

**Comment by jonnylangefeld** [comment]
After lots of testing I'm pretty confident this is working code now. All the checks are succeeding too. I've also updated it to work with the new `attributes` and `server_commands` (vs the old `manufacturer_server_attributes` and `manufacturer_attributes`). @javicalle should we merge it?

File: zhaquirks/tuya/ts110e.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                "Setting brightness to %s",
+                brightness,
+            )
+            return await super().command(
+                TUYA_CUSTOM_LEVEL_COMMAND,
+                TuyaLevelPayload(level=brightness, transtime=0),
+                manufacturer=manufacturer,
+                expect_reply=expect_reply,
+                tsn=tsn,
+            )
```
[the remark this answers is not in the document]
Comment (jonnylangefeld) [reply zhaquirks/tuya/ts110e.py:118]: Hmm that doesn't work because then this Class would also show up as cluster in the dimmers of other devices. `F000LevelControlCluster` has to inherit in some way from `CustomCluster` and `NoManufacturerCluster` does that for us (including setting `manufacturer=foundation.ZCLHeader.NO_MANUFACTURER_ID,`).

[the remark this answers is not in the document]
Comment (javicalle) [reply zhaquirks/tuya/ts110e.py:118]: > this Class would also show up as cluster in the dimmers of other devices

That sentence not make sense to me. A quirk shouldn't modify other devices that don't use the quirk.
Trying to figure out how can be done I believe that some code like this could make the described behaviour:
```python

    # attributes = LevelControl.attributes.copy()
    attributes.update(
        {
            # 0xF000
            TUYA_LEVEL_ATTRIBUTE: ("manufacturer_current_level", t.uint16_t),
        }
    )
```
If not making a copy of parent class attributes, maybe the code is modifying the parent attributes and changes for any device that uses the parent class 🤔 

In any case, your proposed code is fine to me.

File: zhaquirks/tuya/ts110e.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+            # 0xFC02
+            TUYA_BULB_TYPE_ATTRIBUTE: ("bulb_type", TuyaBulbType),
+            # 0xFC03
+            TUYA_MIN_LEVEL_ATTRIBUTE: ("manufacturer_min_level", t.uint16_t),
+            # 0xFC04
+            TUYA_MAX_LEVEL_ATTRIBUTE: ("manufacturer_max_level", t.uint16_t),
+        }
+    )
+
+    # 0xF000 reported values are 10-1000, convert to 0-254
```
[the remark this answers is not in the document]
Comment (javicalle) [reply zhaquirks/tuya/ts110e.py:78]: I agree. We can go ahead with the version as is.

**Review by javicalle — APPROVED** [review]
LGTM

File: zhaquirks/tuya/ts110e.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonnylangefeld) [reply zhaquirks/tuya/ts110e.py:118]: Hmm yeah I don’t know, but it did show up as level control cluster on this new dimmer I bought…
